File: benchmarks/bench_consume.py
"""The `confluent-kafka` half of the consume benchmark.

    python benchmarks/bench_consume.py <bootstrap> <topic> <group> \
        <poll|batch> <batch> <prefetch_ms> <repeats>

Prints the same line shape as every other peer:

    RESULT confluent <mode> <repeat> <messages> <nanoseconds> <stalls> <checksum>

Mirrors the Mojo peer call for call: same config keys, same assign-and-seek
replay, same zero-timeout drain with a stall counter, same first payload byte
summed into a checksum.

**One caveat this file is the right place to record.** `confluent-kafka` ships
a manylinux wheel with its *own* librdkafka bundled -- `ldd` on `cimpl...so`
names `librdkafka-c87086af.so.1` out of `confluent_kafka.libs`, not the
conda-forge library the Mojo, C and Rust peers all load. The version string
matches (2.15.0) but the build does not, so this peer alone is not strictly
comparing binding against binding; the other three are.
"""

import sys
import time

from confluent_kafka import Consumer, KafkaError, TopicPartition, OFFSET_BEGINNING


def config(bootstrap: str, group: str) -> dict:
    return {
        "bootstrap.servers": bootstrap,
        "group.id": group,
        "auto.offset.reset": "earliest",
        "enable.partition.eof": True,
        "enable.auto.commit": False,
        "fetch.max.bytes": 52428800,
        "queued.max.messages.kbytes": 2097151,
        "queued.min.messages": 10000000,
        "fetch.wait.max.ms": 10,
        "log_level": 3,
    }


def is_eof(msg) -> bool:
    err = msg.error()
    return err is not None and err.code() == KafkaError._PARTITION_EOF


def drain(consumer, mode: str, batch: int):
    """One full pass. Returns (seen, checksum, stalls)."""
    seen = 0
    checksum = 0
    stalls = 0
    done = False
    while not done:
        if mode == "poll":
            msg = consumer.poll(0)
            if msg is None:
                stalls += 1
            elif is_eof(msg):
                done = True
            elif msg.error() is not None:
                raise RuntimeError(msg.error())
            else:
                payload = msg.value()
                if payload:
                    checksum += payload[0]
                seen += 1
        elif mode == "batch":
            msgs = consumer.consume(batch, timeout=0)
            if not msgs:
                stalls += 1
            for msg in msgs:
                if is_eof(msg):
                    done = True
                elif msg.error() is not None:
                    raise RuntimeError(msg.error())
                else:
                    payload = msg.value()
                    if payload:
                        checksum += payload[0]
                    seen += 1
        else:
            raise SystemExit(f"mode must be 'poll' or 'batch', got {mode!r}")
        if stalls > 5_000_000:
            done = True
    return seen, checksum, stalls


def main() -> None:
    if len(sys.argv) != 8:
        raise SystemExit(
            "usage: bench_consume.py <bootstrap> <topic> <group> <mode> "
            "<batch> <prefetch_ms> <repeats>"
        )
    bootstrap, topic, group, mode, batch, prefetch_ms, repeats = sys.argv[1:]
    batch, prefetch_ms, repeats = int(batch), int(prefetch_ms), int(repeats)

    consumer = Consumer(config(bootstrap, group))
    # assign, not subscribe -- the repeats seek back to the start and an
    # explicit assignment keeps a rebalance out of the loop.
    consumer.assign([TopicPartition(topic, 0, OFFSET_BEGINNING)])

    for r in range(repeats):
        consumer.seek(TopicPartition(topic, 0, 0))
        time.sleep(prefetch_ms / 1000.0)
        started = time.perf_counter_ns()
        seen, checksum, stalls = drain(consumer, mode, batch)
        elapsed = time.perf_counter_ns() - started
        print(
            f"RESULT confluent {mode} {r} {seen} {elapsed} {stalls} {checksum}",
            flush=True,
        )
    consumer.close()


if __name__ == "__main__":
    main()

File: benchmarks/bench_consume.c
/* The native ceiling: librdkafka driven directly, no binding in the way.
 *
 *   bench_consume_c <bootstrap> <topic> <group> <mode> <batch> <prefetch_ms>
 *                   <repeats>
 *
 * Prints one line per repeat, the same shape every peer prints:
 *
 *   RESULT c <mode> <repeat> <messages> <nanoseconds> <stalls> <checksum>
 *
 * This exists to answer one question the cross-client table cannot: how much
 * of the remaining time is the binding, and how much is librdkafka. Every
 * other peer is measured against it, so it does exactly what they do --
 * same config, same assign/seek replay, same zero-timeout drain, same first
 * payload byte summed into a checksum -- and nothing they do not.
 *
 * `mode` is poll, batch, or batchhdr. The last one is batch plus a
 * rd_kafka_message_headers() call per record, which is what a binding that
 * populates a headers field eagerly has to pay; it is here because that call
 * turned out to cost more than everything else the decode does.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <librdkafka/rdkafka.h>

static long long now_ns(void) {
        struct timespec ts;
        clock_gettime(CLOCK_MONOTONIC, &ts);
        return (long long)ts.tv_sec * 1000000000LL + ts.tv_nsec;
}

static void set(rd_kafka_conf_t *conf, const char *k, const char *v) {
        char err[512];
        if (rd_kafka_conf_set(conf, k, v, err, sizeof(err)) !=
            RD_KAFKA_CONF_OK) {
                fprintf(stderr, "conf %s=%s: %s\n", k, v, err);
                exit(1);
        }
}

int main(int argc, char **argv) {
        if (argc != 8) {
                fprintf(stderr,
                        "usage: %s <bootstrap> <topic> <group> "
                        "<poll|batch|batchhdr> <batch> <prefetch_ms> "
                        "<repeats>\n",
                        argv[0]);
                return 2;
        }
        const char *bootstrap = argv[1], *topic = argv[2], *group = argv[3];
        const char *mode = argv[4];
        int batch = atoi(argv[5]);
        int prefetch_ms = atoi(argv[6]);
        int repeats = atoi(argv[7]);

        rd_kafka_conf_t *conf = rd_kafka_conf_new();
        set(conf, "bootstrap.servers", bootstrap);
        set(conf, "group.id", group);
        set(conf, "auto.offset.reset", "earliest");
        set(conf, "enable.partition.eof", "true");
        set(conf, "enable.auto.commit", "false");
        set(conf, "fetch.max.bytes", "52428800");
        set(conf, "queued.max.messages.kbytes", "2097151");
        set(conf, "queued.min.messages", "10000000");
        set(conf, "fetch.wait.max.ms", "10");
        set(conf, "log_level", "3");

        char err[512];
        rd_kafka_t *rk =
            rd_kafka_new(RD_KAFKA_CONSUMER, conf, err, sizeof(err));
        if (!rk) {
                fprintf(stderr, "rd_kafka_new: %s\n", err);
                return 1;
        }
        rd_kafka_poll_set_consumer(rk);

        rd_kafka_topic_partition_list_t *tpl =
            rd_kafka_topic_partition_list_new(1);
        rd_kafka_topic_partition_list_add(tpl, topic, 0)->offset =
            RD_KAFKA_OFFSET_BEGINNING;
        rd_kafka_assign(rk, tpl);

        rd_kafka_queue_t *queue = rd_kafka_queue_get_consumer(rk);
        rd_kafka_message_t **buf =
            malloc(sizeof(rd_kafka_message_t *) * (size_t)batch);

        for (int r = 0; r < repeats; r++) {
                rd_kafka_topic_partition_list_t *seek =
                    rd_kafka_topic_partition_list_new(1);
                rd_kafka_topic_partition_list_add(seek, topic, 0)->offset = 0;
                rd_kafka_error_t *serr = rd_kafka_seek_partitions(rk, seek, -1);
                if (serr) {
                        fprintf(stderr, "seek: %s\n",
                                rd_kafka_error_string(serr));
                        rd_kafka_error_destroy(serr);
                        return 1;
                }
                rd_kafka_topic_partition_list_destroy(seek);

                struct timespec nap = {prefetch_ms / 1000,
                                       (long)(prefetch_ms % 1000) * 1000000L};
                nanosleep(&nap, NULL);

                long long seen = 0, checksum = 0, stalls = 0;
                int done = 0;
                long long started = now_ns();

                if (!strcmp(mode, "poll")) {
                        while (!done) {
                                rd_kafka_message_t *m =
                                    rd_kafka_consumer_poll(rk, 0);
                                if (!m) {
                                        if (++stalls > 5000000)
                                                done = 1;
                                        continue;
                                }
                                if (m->err ==
                                    RD_KAFKA_RESP_ERR__PARTITION_EOF)
                                        done = 1;
                                else if (!m->err) {
                                        if (m->payload)
                                                checksum +=
                                                    ((unsigned char *)
                                                         m->payload)[0];
                                        seen++;
                                }
                                rd_kafka_message_destroy(m);
                        }
                } else {
                        int want_hdrs = !strcmp(mode, "batchhdr");
                        while (!done) {
                                ssize_t n = rd_kafka_consume_batch_queue(
                                    queue, 0, buf, (size_t)batch);
                                if (n < 0) {
                                        fprintf(stderr, "batch: %s\n",
                                                rd_kafka_err2str(
                                                    rd_kafka_last_error()));
                                        return 1;
                                }
                                if (n == 0) {
                                        if (++stalls > 5000000)
                                                done = 1;
                                        continue;
                                }
                                for (ssize_t i = 0; i < n; i++) {
                                        rd_kafka_message_t *m = buf[i];
                                        if (m->err ==
                                            RD_KAFKA_RESP_ERR__PARTITION_EOF)
                                                done = 1;
                                        else if (!m->err) {
                                                if (m->payload)
                                                        checksum +=
                                                            ((unsigned char *)
                                                                 m->payload)[0];
                                                if (want_hdrs) {
                                                        rd_kafka_headers_t *h;
                                                        checksum +=
                                                            rd_kafka_message_headers(
                                                                m, &h) &
                                                            1;
                                                }
                                                seen++;
                                        }
                                        rd_kafka_message_destroy(m);
                                }
                        }
                }
                long long elapsed = now_ns() - started;
                printf("RESULT c %s %d %lld %lld %lld %lld\n", mode, r, seen,
                       elapsed, stalls, checksum);
                fflush(stdout);
        }

        free(buf);
        rd_kafka_queue_destroy(queue);
        rd_kafka_consumer_close(rk);
        rd_kafka_topic_partition_list_destroy(tpl);
        rd_kafka_destroy(rk);
        return 0;
}
